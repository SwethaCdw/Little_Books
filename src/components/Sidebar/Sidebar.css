.sidebar-container {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: 16%;
  background-color: white;
  border: none;
  height: 100vh;
}
.sidebar-container.dark-mode {
  background-color: #021929;
  color: white;
}
.sidebar-container.dark-mode .view-members, .sidebar-container.dark-mode .app-theme {
  color: white !important;
}
.sidebar-container.dark-mode .view-members:hover, .sidebar-container.dark-mode .app-theme:hover {
  color: gray !important;
}
.sidebar-container .logo {
  font-size: 28px;
  font-family: "Roboto Slab", serif;
  font-weight: 400;
  padding: 50px 20px;
  text-align: center;
}
.sidebar-container .logo span:first-child {
  color: #b354b7;
}
.sidebar-container .logo span:nth-child(2) {
  font-weight: 500;
}
.sidebar-container .filter-items {
  list-style: none;
}
.sidebar-container .filter-items .filter-heading {
  padding-bottom: 10px;
}
.sidebar-container .filter-items li {
  display: flex;
  gap: 10px;
  padding-bottom: 10px;
}
.sidebar-container .filter-items li input {
  accent-color: #b354b7;
  width: 17px;
  cursor: pointer;
}
.sidebar-container .filter-items li .filter-item {
  font-size: 15px;
}
.sidebar-container .buttons {
  margin-bottom: 45px;
}
.sidebar-container .buttons .view-members, .sidebar-container .buttons .app-theme {
  font-size: 15px;
  border: none;
  background-color: transparent;
  color: black;
  padding-bottom: 10px;
  cursor: pointer;
}
.sidebar-container .buttons .view-members:hover, .sidebar-container .buttons .app-theme:hover {
  color: gray;
}/*# sourceMappingURL=Sidebar.css.map */