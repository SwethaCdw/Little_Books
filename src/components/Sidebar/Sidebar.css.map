{"version":3,"sources":["Sidebar.scss","Sidebar.css"],"names":[],"mappings":"AAAA;EACI,aAAA;EACA,sBAAA;EACA,8BAAA;EACA,mBAAA;EACA,UAAA;EACA,uBAAA;EACA,YAAA;EACA,aAAA;ACCJ;ADAI;EACI,yBAAA;EACA,YAAA;ACER;ADDQ;EACI,uBAAA;ACGZ;ADFY;EACI,sBAAA;ACIhB;ADAI;EACI,eAAA;EACA,iCAAA;EACA,gBAAA;EACA,kBAAA;EACA,kBAAA;ACER;ADAQ;EACI,cAAA;ACEZ;ADAQ;EACI,gBAAA;ACEZ;ADEI;EACE,gBAAA;ACAN;ADCM;EACE,oBAAA;ACCR;ADCM;EACE,aAAA;EACA,SAAA;EACA,oBAAA;ACCR;ADAQ;EACI,qBAAA;EACA,WAAA;EACA,eAAA;ACEZ;ADAQ;EACI,eAAA;ACEZ;ADEI;EACI,mBAAA;ACAR;ADCQ;EACI,eAAA;EACA,YAAA;EACA,6BAAA;EACA,YAAA;EACA,oBAAA;EACA,eAAA;ACCZ;ADAY;EACI,WAAA;ACEhB","file":"Sidebar.css"}