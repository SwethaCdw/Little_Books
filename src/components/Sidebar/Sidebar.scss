.sidebar-container {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
    width: 16%;
    background-color: white;
    border: none;
    height: 100vh;
    &.dark-mode {
        background-color: #021929;
        color: white;
        .view-members, .app-theme {
            color: white !important;
            &:hover {
                color: gray !important;
            }
        }
    }
    .logo {
        font-size: 28px;
        font-family: "Roboto Slab", serif;
        font-weight: 400;
        padding: 50px 20px;
        text-align: center;

        span:first-child {
            color: #b354b7;
        }
        span:nth-child(2) {
            font-weight: 500;
        }

    }
    .filter-items {
      list-style: none;
      .filter-heading {
        padding-bottom: 10px;
      }
      li {
        display: flex;
        gap: 10px;
        padding-bottom: 10px;
        input {
            accent-color: #b354b7;
            width: 17px;
            cursor: pointer;
        }
        .filter-item {
            font-size: 15px;
        }
      }
    }
    .buttons {
        margin-bottom: 45px;
        .view-members, .app-theme {
            font-size: 15px;
            border: none;
            background-color: transparent;
            color: black;
            padding-bottom: 10px;
            cursor: pointer;
            &:hover {
                color: gray;
            }
        }
    }
    
}