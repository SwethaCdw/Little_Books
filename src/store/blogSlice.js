import { createSlice } from '@reduxjs/toolkit';
import { BLOG_TYPE } from '../constants/common-constants';
import { setItemInLocalStorage } from '../utils/local-storage-utils';

const blogSlice = createSlice({
  name: 'blog',
  initialState: {
    blogsData: [],
    searchTerm: '',
    selectedBlog: null,
    imageLoaded: false,
    showNewBlogModal: false,
    selectedFilters: BLOG_TYPE
  },
  reducers: {
    setBlogsData(state, action) {
      state.blogsData = action.payload;
      setItemInLocalStorage('blogs', JSON.stringify(action.payload));
    },
    setSearchTerm(state, action) {
      state.searchTerm = action.payload;
    },
    setSelectedBlog(state, action) {
      state.selectedBlog = action.payload;
    },
    setImageLoaded(state, action) {
      state.imageLoaded = action.payload;
    },
    addNewBlog: (state, action) => {
      state.blogsData.push(action.payload);
    },
    setShowNewBlogModal: (state, action) => {
      state.showNewBlogModal = action.payload;
    },
    updateSelectedBlog: (state, action) => {
      const updatedBlog = action.payload;
      const index = state.blogsData.findIndex(blog => blog.photo === updatedBlog.photo);
      if (index !== -1) {
        state.blogsData[index] = updatedBlog;
        setItemInLocalStorage('blogs', JSON.stringify(state.blogsData));

      }
      state.selectedBlog = updatedBlog;

    }, 
    updateSelectedFilter : (state, action) => {
      state.selectedFilters = action.payload;
    }
  }
});

export const { setBlogsData, setSearchTerm, setSelectedBlog, setImageLoaded, addNewBlog, setShowNewBlogModal, updateSelectedBlog, updateSelectedFilter} = blogSlice.actions;

export default blogSlice.reducer;
