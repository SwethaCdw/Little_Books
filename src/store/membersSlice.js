import { createSlice } from '@reduxjs/toolkit';

const membersSlice = createSlice({
  name: 'members',
  initialState: {
    openModal: false,
    membersData: null,
  },
  reducers: {
    setOpenModal(state, action) {
      state.openModal = action.payload;
    },
    setMembersData(state, action) {
      state.membersData = action.payload;
    },
   
  }
});

export const { setOpenModal, setMembersData } = membersSlice.actions;

export default membersSlice.reducer;
