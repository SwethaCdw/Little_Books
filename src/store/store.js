import { configureStore } from '@reduxjs/toolkit';
import blogReducer from './blogSlice';
import membersReducer from './membersSlice';
import themeReducer from './themeSlice';

const store = configureStore({
  reducer: {
    blog: blogReducer,
    members: membersReducer,
    theme: themeReducer
  }
});

export default store;
