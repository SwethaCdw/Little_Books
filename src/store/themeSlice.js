import { createSlice } from '@reduxjs/toolkit';
import { getItemFromLocalStorage, setItemInLocalStorage } from '../utils/local-storage-utils';

const initialState = {
  darkMode: getItemFromLocalStorage('darkMode'),
};

const themeSlice = createSlice({
  name: 'theme',
  initialState,
  reducers: {
    toggleDarkMode: state => {
      state.darkMode = !state.darkMode;
      setItemInLocalStorage('darkMode', state.darkMode);
    },
  },
});

export const { toggleDarkMode } = themeSlice.actions;

export default themeSlice.reducer;