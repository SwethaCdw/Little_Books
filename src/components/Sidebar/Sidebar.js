
import React from 'react';
import './Sidebar.css';
import { useDispatch, useSelector } from 'react-redux';
import { setOpenModal } from '../../store/membersSlice';
import Members from '../Members/Members';
import { toggleDarkMode } from '../../store/themeSlice';
import { updateSelectedFilter } from '../../store/blogSlice';

const Sidebar = () => {
  const dispatch = useDispatch();
  const openModal = useSelector((state) => state.members.openModal)
  const darkMode = useSelector((state) => state.theme.darkMode);
  const selectedFilters = useSelector((state) => state.blog.selectedFilters);
  const blogsData = useSelector((state) => state.blog.blogsData);
  const types = [...new Set(blogsData.map(blog => blog.type))];

  const openMembersModal = () => {
    dispatch(setOpenModal(true));
  }

  const handleModalClose = () => {
    dispatch(setOpenModal(false));
  };

  const handleThemeToggle = () => {
    dispatch(toggleDarkMode()); 
  };

  const handleCheckboxChange = (type) => {
    const updatedTypes = selectedFilters.includes(type)
      ? selectedFilters.filter((selectedType) => selectedType !== type)
      : [...selectedFilters, type];
      dispatch(updateSelectedFilter(updatedTypes));
  };


  return (
    <>
   
   <div className={`sidebar-container ${darkMode ? 'dark-mode' : ''}`}>
        <p className='logo'><span>Little</span> <span>Book</span></p>
        <div >
          <ul className='filter-items'>
            <h4 className='filter-heading'>FILTER</h4>
          {types.map((type) => (
            <label key={type}>
              <li >
                  <input
                    type='checkbox'
                    checked={selectedFilters.includes(type)}
                    onChange={() => handleCheckboxChange(type)}
                  />
                  <p className='filter-item'>{type} Blogs</p>
              </li>
              </label>
            ))}
          </ul>

        </div>
        <div className='buttons'>
            <p className='view-members' onClick={openMembersModal}>View Members</p>
            <p className='app-theme' onClick={handleThemeToggle}>
              {darkMode ? 'Switch to Light Mode' : 'Switch to Dark Mode'}
            </p>
        </div>
    </div>
    {openModal && <Members onClose={handleModalClose}/>}
    </>
  );
};

export default Sidebar;