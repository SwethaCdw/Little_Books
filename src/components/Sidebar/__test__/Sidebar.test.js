import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Sidebar from '../Sidebar';
import { setOpenModal } from '../../../store/membersSlice';
import { toggleDarkMode } from '../../../store/themeSlice';

const mockStore = configureStore([]);

describe('Sidebar component', () => {

  let store;

  beforeEach(() => {
    store = mockStore({
      members: { openModal: false },
      theme: { darkMode: false },
      blog: { selectedFilters: [], blogsData: [{ type: 'Technology' }, { type: 'Travel' }] }
    });
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<Provider store={store}><Sidebar /></Provider>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders sidebar component', () => {
    render(
      <Provider store={store}>
        <Sidebar />
      </Provider>
    );

    const littleSpan = screen.getByText('Little');
    const bookSpan = screen.getByText('Book');
  
    expect(littleSpan).toBeInTheDocument();
    expect(bookSpan).toBeInTheDocument();
  });

  test('opens members modal when clicked on "View Members"', () => {
    render(
      <Provider store={store}>
        <Sidebar />
      </Provider>
    );

    fireEvent.click(screen.getByText('View Members'));

    expect(store.getActions()).toContainEqual(setOpenModal(true));
  });

  test('toggles theme when clicked on "Switch to Dark Mode"', () => {
    render(
      <Provider store={store}>
        <Sidebar />
      </Provider>
    );

    fireEvent.click(screen.getByText('Switch to Dark Mode'));

    expect(store.getActions()).toContainEqual(toggleDarkMode());
  });

  test('updates selected filter when checkbox is clicked', () => {
    render(
      <Provider store={store}>
        <Sidebar />
      </Provider>
    );
    const technologyCheckbox = screen.getByLabelText('Technology Blogs');

    fireEvent.click(technologyCheckbox);
  
    expect(technologyCheckbox.checked).toBe(false);

  });
});
