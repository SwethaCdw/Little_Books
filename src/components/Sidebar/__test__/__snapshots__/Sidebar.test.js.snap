// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sidebar-container "
  >
    <p
      class="logo"
    >
      <span>
        Little
      </span>
       
      <span>
        Book
      </span>
    </p>
    <div>
      <ul
        class="filter-items"
      >
        <h4
          class="filter-heading"
        >
          FILTER
        </h4>
        <label>
          <li>
            <input
              type="checkbox"
            />
            <p
              class="filter-item"
            >
              Technology Blogs
            </p>
          </li>
        </label>
        <label>
          <li>
            <input
              type="checkbox"
            />
            <p
              class="filter-item"
            >
              Travel Blogs
            </p>
          </li>
        </label>
      </ul>
    </div>
    <div
      class="buttons"
    >
      <p
        class="view-members"
      >
        View Members
      </p>
      <p
        class="app-theme"
      >
        Switch to Dark Mode
      </p>
    </div>
  </div>
</DocumentFragment>
`;
